File: frontend/coordinator-android/app/src/main/java/ug/qaat/coordinator/net/AuthClient.kt
package ug.qaat.coordinator.net

import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import org.json.JSONObject

/**
 * Coordinator login — mirrors the PWA exactly:
 *   GET  /api/v1/auth/tenant-lookup?email=  → { tenant_id }
 *   POST /api/v1/auth/login {email,password,totp_code,tenant_id}
 *        → { access_token, jti, role, user_id, expires_in, device_binding_key }
 *
 * device_binding_key is the secret the Sealer needs (HKDF → AES/HMAC). Keep it only
 * in memory / Android Keystore — never log it.
 */
class AuthClient {
    private val http = Net.client()
    private val base = Net.baseUrl

    private companion object {
        const val NON_JSON = "The server returned a web page instead of data — it's likely waking up (it sleeps when idle). Wait ~30s and tap Sign in again."
        fun String.looksJson() = trimStart().let { it.startsWith("{") || it.startsWith("[") }
        fun httpErr(code: Int): String = when (code) {
            in 500..599 -> "The server is waking up or busy (HTTP $code). Wait ~30s and try again."
            401, 403 -> "Incorrect email or password."
            404 -> "No account found for that email."
            else -> "Sign in failed (HTTP $code)."
        }
    }

    data class Result(
        val token: String, val jti: String?, val role: String,
        val userId: String, val tenantId: String, val deviceBindingKey: String?,
        val fullName: String, val title: String, val registrationNo: String,
        // Unified app-login extras: the student's reg-number / the lecturer's staff-id, and the
        // institution slug the user signed in with (kept for later reg-scoped online calls).
        val studentId: String = "", val staffId: String = "", val org: String = "",
    )

    /**
     * Unified KIU QAAT sign-in for EVERY role. identifier = email (staff/coordinator) OR reg-number
     * (student) OR staff-id (lecturer); org = institution id/domain. Hits the gateway's app-login,
     * which resolves the account, reuses the real /auth/login, and augments with student_id/staff_id.
     * @return Result on success; null with [onMfaRequired] if the server wants a TOTP code.
     */
    suspend fun appLogin(identifier: String, password: String, org: String, totp: String?, onMfaRequired: () -> Unit): Result? {
        val body = JSONObject()
            .put("identifier", identifier).put("password", password)
            .put("org", org).put("totp_code", totp ?: "")
        var text = ""; var status = 0
        for (attempt in 1..6) {
            val r = http.post("$base/api/v1/auth/app-login") {
                contentType(ContentType.Application.Json); setBody(body.toString())
            }
            text = r.bodyAsText(); status = r.status.value
            if (text.looksJson() || status < 500) break
            if (attempt < 6) kotlinx.coroutines.delay(6000)
        }
        if (!text.looksJson())
            throw IllegalStateException(if (status in 200..299) NON_JSON else httpErr(status))
        val j = JSONObject(text)
        if (status == 403 && j.optString("error") == "MFA_REQUIRED") { onMfaRequired(); return null }
        require(status in 200..299) { j.optString("message").ifBlank { httpErr(status) } }
        return Result(
            token = j.getString("access_token"),
            jti = j.optString("jti").takeIf { it.isNotEmpty() },
            role = j.optString("role", ""),
            userId = j.getString("user_id"),
            tenantId = j.optString("tenant_id", ""),
            deviceBindingKey = j.optString("device_binding_key").takeIf { it.isNotEmpty() },
            fullName = j.optString("full_name", ""),
            title = j.optString("title", ""),
            registrationNo = j.optString("registration_number", ""),
            studentId = j.optString("student_id", ""),
            staffId = j.optString("staff_id", ""),
            org = j.optString("org", org),
        )
    }

    /** Change the signed-in coordinator's password. @return error message, or null on success. */
    suspend fun changePassword(token: String, current: String, newPassword: String): String? {
        val r = http.post("$base/api/v1/auth/change-password") {
            header("Authorization", "Bearer $token"); contentType(ContentType.Application.Json)
            setBody(JSONObject().put("current_password", current).put("new_password", newPassword).toString())
        }
        if (r.status.value in 200..299) return null
        return runCatching { JSONObject(r.bodyAsText()).optString("message", "Could not change password") }
            .getOrDefault("Could not change password")
    }

    suspend fun tenantLookup(email: String): String {
        val r = http.get("$base/api/v1/auth/tenant-lookup") { url { parameters.append("email", email) } }
        val text = r.bodyAsText()
        if (!r.status.isSuccess())
            throw IllegalStateException(if (text.looksJson()) JSONObject(text).optString("message").ifBlank { httpErr(r.status.value) } else httpErr(r.status.value))
        if (!text.looksJson()) throw IllegalStateException(NON_JSON)
        return JSONObject(text).getString("tenant_id")
    }

    /** @return Result on success; null with [mfaRequired]=true if the server wants a TOTP code. */
    suspend fun login(email: String, password: String, totp: String?, onMfaRequired: () -> Unit): Result? {
        val tenantId = tenantLookup(email)
        val body = JSONObject()
            .put("email", email).put("password", password)
            .put("totp_code", totp ?: "").put("tenant_id", tenantId)
        // Cold-start resilience: a sleeping free-tier auth-service answers with a 502/HTML page while
        // it wakes (~15–60s). Auto-retry through the wake so a FRESH login (e.g. on a phone with no
        // saved session) just completes after a short wait, instead of showing "waking up" over and
        // over. A JSON body — even a 401 — means the service actually answered, so we stop retrying.
        var text = ""; var status = 0
        for (attempt in 1..6) {
            val r = http.post("$base/api/v1/auth/login") {
                contentType(ContentType.Application.Json); setBody(body.toString())
            }
            text = r.bodyAsText(); status = r.status.value
            if (text.looksJson() || status < 500) break   // service answered
            if (attempt < 6) kotlinx.coroutines.delay(6000) // still waking — wait, then retry
        }
        if (!text.looksJson())
            throw IllegalStateException(if (status in 200..299) NON_JSON else httpErr(status))
        val j = JSONObject(text)
        if (status == 403 && j.optString("error") == "MFA_REQUIRED") { onMfaRequired(); return null }
        require(status in 200..299) { j.optString("message").ifBlank { httpErr(status) } }
        return Result(
            token = j.getString("access_token"),
            jti = j.optString("jti").takeIf { it.isNotEmpty() },
            role = j.optString("role", "COORDINATOR"),
            userId = j.getString("user_id"),
            tenantId = tenantId,
            deviceBindingKey = j.optString("device_binding_key").takeIf { it.isNotEmpty() },
            fullName = j.optString("full_name", ""),
            title = j.optString("title", ""),
            registrationNo = j.optString("registration_number", ""),
        )
    }
}

File: frontend/coordinator-flutter/lib/net/auth_client.dart
import '../net/net.dart';

/// Sign-in response from `POST /api/v1/auth/app-login` — the UNIFIED KIU QAAT
/// login for every role. Mirrors the gateway's auth_applogin.go augmentation:
/// on top of the auth-service `loginResponse` it carries the student's
/// registration number and the lecturer's/patroller's staff id, which the
/// coordinator-flutter port routes by role.
class LoginResult {
  const LoginResult({
    required this.token,
    required this.role,
    required this.userId,
    required this.tenantId,
    this.fullName = '',
    this.title = '',
    this.registrationNo = '',
    this.studentId = '',
    this.staffId = '',
    this.deviceBindingKey,
    this.forcePasswordChange = false,
  });

  factory LoginResult.fromJson(Map<String, dynamic> j) {
    final bindingKey = (j['device_binding_key'] as String?) ?? '';
    return LoginResult(
      token: (j['access_token'] ?? '') as String,
      role: (j['role'] ?? '') as String,
      userId: (j['user_id'] ?? '') as String,
      tenantId: (j['tenant_id'] ?? '') as String,
      fullName: (j['full_name'] ?? '') as String,
      title: (j['title'] ?? '') as String,
      registrationNo: (j['registration_number'] ?? '') as String,
      studentId: (j['student_id'] ?? '') as String,
      staffId: (j['staff_id'] ?? '') as String,
      deviceBindingKey: bindingKey.isEmpty ? null : bindingKey,
      forcePasswordChange: (j['force_password_change'] ?? false) as bool,
    );
  }

  /// The server field names, so a saved session round-trips through the same API
  /// shape and a future backend that returns more fields keeps working.
  Map<String, dynamic> toJson() => {
        'access_token': token,
        'role': role,
        'user_id': userId,
        'tenant_id': tenantId,
        'full_name': fullName,
        'title': title,
        'registration_number': registrationNo,
        'student_id': studentId,
        'staff_id': staffId,
        if (deviceBindingKey != null) 'device_binding_key': deviceBindingKey,
        'force_password_change': forcePasswordChange,
      };

  final String token;
  final String role;
  final String userId;
  final String tenantId;
  final String fullName;
  final String title;
  final String registrationNo;
  final String studentId;
  final String staffId;

  /// Server-issued coordinator device binding key (`device_binding_key`) — the
  /// secret the session-package Sealer derives its AES/HMAC keys from. Keep it in
  /// memory only; never log it.
  final String? deviceBindingKey;

  /// True when the account still signs in with its seeded default password; the
  /// app must force a change before showing any role UI.
  final bool forcePasswordChange;

  /// Best display name: the credential's full name, else the identifier's local part.
  String displayName(String identifier) =>
      fullName.isNotEmpty ? fullName : identifier.split('@').first;
}

/// Login / credential helpers hitting the gateway's public auth routes.
class AuthClient {
  const AuthClient();

  /// POST /api/v1/auth/app-login  (public, rate-limited, 429 back-off in `Net`).
  ///
  /// [identifier] = email (staff/coordinator) OR registration number (student) OR
  /// staff ID (lecturer). Returns the signed-in [LoginResult]; on the server's
  /// `403 {error: MFA_REQUIRED}` it returns null and the caller should show the
  /// TOTP field and submit again with [totp] populated.
  Future<LoginResult?> appLogin(
    String identifier,
    String password, {
    String? totp,
  }) async {
    try {
      final j = await Net.postJson(
        '/api/v1/auth/app-login',
        body: {
          'identifier': identifier.trim(),
          'password': password,
          if (totp != null && totp.isNotEmpty) 'totp_code': totp.trim(),
        },
      ) as Map<String, dynamic>;
      return LoginResult.fromJson(j);
    } on NetFailure catch (e) {
      if (e.isMfaRequired) return null;
      rethrow;
    }
  }

  /// POST /api/v1/auth/change-password — the signed-in user changes their own
  /// password. Returns the server's refusal message, or null on success.
  Future<String?> changePassword(
    String token,
    String current,
    String newPassword,
  ) async {
    try {
      await Net.postJson(
        '/api/v1/auth/change-password',
        token: token,
        body: {'current_password': current, 'new_password': newPassword},
      );
      return null;
    } on NetFailure catch (e) {
      return e.message;
    }
  }
}